pkg/server/http/middleware/request_id.go

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package middleware
```

Comment:
cc @yashrsharma44

Reply:
Looks great, however, do we want to correlate the trace-id with request-id? Maybe my idea would be suitable for another PR for correlating them.
Otherwise looks great :100:

Reply:
We can try correlate or use a single id for both. We need to check how actually different middleware interacts and what happens if tracing is not enabled. For the scope of this PR, I want to keep track of execution of async forward requests in the scope a single request.

Reply:
Aligning everything with the W3C trace context would probably work, but doesn't block this I would say.